boundedint/int256.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2020 The Swarm Authors
+// This file is part of the Swarm library.
+//
+// The Swarm library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Swarm library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
+
+package boundedint
+
+import (
+	"fmt"
+	"io"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+// Int256 represents an signed integer of 256 bits
+type Int256 struct {
+	value big.Int
+}
+
+var minInt256 = new(big.Int).Mul(big.NewInt(-1), new(big.Int).Exp(big.NewInt(2), big.NewInt(255), nil)) // -(2^255)
+var maxInt256 = new(big.Int).Sub(new(big.Int).Exp(big.NewInt(2), big.NewInt(255), nil), big.NewInt(1))  // 2^255 - 1
+
+// NewInt256 creates a Int256 struct with an initial underlying value of zero
+func NewInt256() *Int256 {
+	u := new(Int256)
+	u.value = *new(big.Int).Set(big.NewInt(0))
```

Comment:
the reason why `value` is stored as a plain `big.Int` (rather than a pointer) is to avoid an `Int256` variable with a `nil` underlying value (instead of `0`). see [this discussion](https://github.com/ethersphere/swarm/pull/2063#issuecomment-571280885) from the first PR for this issue.

regarding the de-referencing: it appears without it, the `value` field is subject to modification from outside of the `Set` function. please see [this comment](https://github.com/ethersphere/swarm/pull/2063#discussion_r364398878) from that same PR. 

finally: can you please explain what you mean by this part of your comment?
> you also wrote a set of getters and setters that allow you to choose if to allow the caller to work on a copy or on the actual pointer value.

how would you go about "working on a copy or on the actual pointer value"? could you please give an example?

Reply:
> the reason why value is stored as a plain big.Int (rather than a pointer) is to avoid an Int256 variable with a nil underlying value (instead of 0). see this discussion from the first PR for this issue.

I'm still not sure I understand why you need all this jumping through hoops. If the type instance was initialized with a `nil` value it means that it is a developer mistake and that proper testing should uncover the mistake. Use a constructor, like you do here, and set the instance value to be something, and not nil. Where's the problem?

The comment links don't work btw so:
> the value field is subject to modification from outside of the Set function

This is true only if you export the value field no? If immutability is important here then the getter can return `big.Int` (and not `*big.Int`) and when returning the value you would dereference the pointer.

It could be that you would actually need to allocate and copy the value manually to return a copy and not the underlying value, but i find this more clear than these tactics.

Also, TBH, I don't like the fact that the type has a `Set` method that overrides the value, it would be nicer that if you want to initialize the int to a certain value - you would supply that value in the constructor. it should not be possible to mutate the value afterwards by overriding (unless you envision pooling the instances at some point), except for using the arithmetic functions on the type itself.

Reply:
will properly reply to this comment shortly, but regarding the links: 

you need to give the page a few seconds after loading for it to take you to the comment.

i've verified the links in different browsers just in case

Reply:
> I'm still not sure I understand why you need all this jumping through hoops. If the type instance was initialized with a `nil` value it means that it is a developer mistake and that proper testing should uncover the mistake. Use a constructor, like you do here, and set the instance value to be something, and not nil. Where's the problem?

we want this new type to behave as similarly as possible with the variables that now use it (in comparison to their previous type).

previously, our zero-value for the relevant variables was `0`. therefore, it's better to have a `big.Int` with a value of `0` instead of a `nil` pointer in its stead.

> Also, TBH, I don't like the fact that the type has a `Set` method that overrides the value, it would be nicer that if you want to initialize the int to a certain value - you would supply that value in the constructor. it should not be possible to mutate the value afterwards by overriding (unless you envision pooling the instances at some point), except for using the arithmetic functions on the type itself.

i agree